### ISSUE TITLE
Detection time on Jetson TX1

### ISSUE THREAD

**Issue by forever3000** [weiliu89/caffe#166]
Hi @weiliu89 
I have just tested SSD with pre-trained reduced VGG network model on Jetson TX1 platform. The fps got only 5-6, same as Jetson TK1 performance. Do you have any idea?
Thanks,

**Comment by clarkong** [comment]
have you maximized CPU performance or fixed GPU frequence?

**Comment by forever3000** [comment]
@clarkong 
Maybe I tried maximum script but something can not applied. This is performance on 64-bit Jetson TX1. I also continue testing on 32-bit to see what difference.
Thanks

**Comment by forever3000** [comment]
hi,
Update : 
1. After applied maximized CPU performance and fixed GPU frequence, fps is 7. It seem like 30-40% faster
2. I tried 32-bit version on Jetson TX1, look like same.
Thanks

**Comment by Josca** [comment]
@forever3000 Hi, did you use SSD300 or SSD500?

**Comment by forever3000** [comment]
@Josca 
I'm using SSD 300x300. I want to get about 13-15 fps with only 4-5 classes on TX1 so  I also tried some base network models (VGG-16 reduced, RESNET 18 layer) but speed got a little difference. May you have some advices.
Thanks

**Comment by Josca** [comment]
I'm quite new at using Caffe and SSD, but I'm very interested in it. I would like to use ZF as a base SSD net, but I don't know how to replace VGG net currently. I think it's possible to download ZF model from [here](https://github.com/ShaoqingRen/faster_rcnn). Good luck.

**Comment by xuyongxi** [comment]
@forever3000,  how many fps  when you use SSD 300X300 on TK1 ?

**Comment by forever3000** [comment]
@xuyongxi sorry for late reply. I got about 5-6 fps after applying maximum performance script.

**Comment by hengck23** [comment]
@forever3000 
For you TX1 timing, is it for batch=1?

**Comment by forever3000** [comment]
@hengck23 ,
In detection phase, I think batch always 1. Please correct if I'm wrong.
Thanks

**Comment by hengck23** [comment]
Googlenet bn runs at about 7 fps on tx1 for modified size of 320x240 for 20 classes. However i am facing some accuracy issue

**Comment by forever3000** [comment]
@hengck23 ,
You replace Reduced VGG with googlenet? Can you share your network and accuracy

**Comment by hengck23** [comment]
@forever3000
Please also refer to #issue 196. My mAP is about 63% but it is not optimum i think.
https://github.com/weiliu89/caffe/issues/196

**Comment by karthikmswamy** [comment]
@forever3000 Did you use the Python script for the timing? Are you looking at the FPS numbers on the window when running python_webcam.py? It uses the clock() call which times the CPU. I modified the C++ in ssd_detect.cpp using gettimeofday() and a forward pass on TK1 takes close to 630 ms.  Can you confirm this? Thanks.

**Comment by forever3000** [comment]
@karthikmswamy 
I'm using ssd_pascal_video.py to play one video and looking at the FPS number on the windows. I think python_webcam.py have same. Do you concern about the FPS calculation in python script ?

**Comment by karthikmswamy** [comment]
@forever3000 Thanks for the confirmation. Yes, the python script uses clock() which seems quite unreliable. The C++ gettimeofday() seems to be fair. Forward pass on my TK1 requires close to 630 ms and I'm able to reduce this to 587 ms by modifying some network parameters, but not any better.
@weiliu89 Could you confirm if clock() is good for timing forward pass? Some tutorials suggest default_timer() like this one: https://github.com/harrism/numbapro_examples/blob/master/mandelbrot_numbapro.ipynb

**Comment by forever3000** [comment]
@karthikmswamy 
Are you running on video or just one image. Can you share your C++ modification and how to run this?
Thanks

**Comment by karthikmswamy** [comment]
@forever3000 I run the webcam part. In any case, the forward pass time is a constant whether we run a single file or a video as long as the batch size is 1. I've provided a sample timing code in the snippet below. 
`static struct timeval tval_bf_pred, tval_af_pred, tval_res_pred;
gettimeofday(&tval_bf_pred, NULL);
detections = detector.Detect(img); // Code from the repo, all others wrap around this.
gettimeofday(&tval_af_pred, NULL);
timersub(&tval_af_pred, &tval_bf_pred, &tval_res_pred);
net_pred_time += (tval_res_pred.tv_sec + tval_res_pred.tv_usec/1000000.0);
`

**Comment by forever3000** [comment]
@karthikmswamy ,
Thank, I will try to confirm. How about your FPS looking on windows? Do you have plan to test on TX1?

**Comment by karthikmswamy** [comment]
@forever3000 Thanks. Please update the FPS you get on TX1. I do hope to understand which timing is reliable for calculating the FPS. Thanks.

**Comment by forever3000** [comment]
@karthikmswamy ,
I try with my PC first (GTX960), I replaced clock() with gettimeofday in VisualizeBBox function (bbox_util.cpp) then tested video detection. The FPS has been changed from 20 -> 24. Update to you.

**Comment by weiliu89** [comment]
@karthikmswamy @forever3000 I also realized that clock() is not accurate. It turns out that caffe has a [timer](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/util/benchmark.cpp#L121), which uses boost::posix_time. I think it is used in `./build/tools/caffe time` as well. Replacing clock() with CPUTimer at [here](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/util/bbox_util.cpp#L1142) might estimate the FPS better when running on video or webcam. Correct me if I am wrong.

In the paper, I estimate the FPS by dividing the total time by the total number of images in a LMDB file.

**Comment by karthikmswamy** [comment]
@weiliu89 Thanks for your update. The Caffe built-in timer is good, not quite which one is reliable. Could you let me know if you use the Caffe CPUTimer to perform your forward pass tests? As @forever3000 mentioned above, it looks like replacing with gettimeofday is useful. I'm waiting for his update on TX1 forward pass timing.

**Comment by forever3000** [comment]
@karthikmswamy 
I also updated forward pass timing on TX1 on the previous comment 
**I also tested on TX1, the FPS got only 2-3 with gettimeofday function. You can refer this #196 for faster speed . I will test it soon.**

**Comment by gilshm** [comment]
Any luck making SSD faster on TX1? I also get 3FPS.

**Comment by karthikmswamy** [comment]
@gilshm I managed to run it faster by re-compiling OpenCV with smaller buffer size. This might not be the ideal way to improve speed but it worked. Try out the latest code in this repo. It works faster than the earlier version.

**Comment by tamnicer** [comment]
@forever3000 
hi! i saw that you ran ssd (for 4-5 class) on TX1 board and got 12-13 fps.

did you change only the number of class number by fine tuning VGG16 model?
(when fine tuning, did you change just the name of _mbox_conf to _mbox_conf_else ??)
please share your experiance!
